File: .github/scripts/test_imports.py
#!/usr/bin/env python3
"""
Test Imports Script
Validates that all Python files are properly wired with global awareness and accessibility
Tests for silent failures in imports and functions
"""

import sys
import os
import importlib.util
from pathlib import Path

# Constants
ERROR_MESSAGE_MAX_LENGTH = 100


def test_file_import(file_path: str, base_dir: str) -> tuple[bool, str]:
    """
    Test if a Python file can be imported without errors
    
    Args:
        file_path: Path to Python file relative to base_dir
        base_dir: Base directory containing the Python files
        
    Returns:
        tuple: (success: bool, message: str)
    """
    try:
        full_path = os.path.join(base_dir, file_path)
        
        if not os.path.exists(full_path):
            return False, f"File not found: {file_path}"
        
        # Load and import the module
        module_name = file_path.replace('/', '.').replace('.py', '')
        spec = importlib.util.spec_from_file_location(module_name, full_path)
        
        if spec is None or spec.loader is None:
            return False, f"Cannot create module spec for {file_path}"
        
        module = importlib.util.module_from_spec(spec)
        
        # Add parent directory to sys.path for relative imports
        parent_dir = os.path.dirname(full_path)
        if parent_dir not in sys.path:
            sys.path.insert(0, parent_dir)
        
        # Execute the module
        spec.loader.exec_module(module)
        
        return True, f"Successfully imported {file_path}"
        
    except Exception as e:
        return False, f"Error importing {file_path}: {str(e)[:ERROR_MESSAGE_MAX_LENGTH]}"


def test_utils_functions():
    """Test utils.py functions are accessible"""
    try:
        import utils
        
        required_functions = ['load_trades_index', 'load_account_config']
        
        for func_name in required_functions:
            if not hasattr(utils, func_name):
                return False, f"Missing function: utils.{func_name}"
            if not callable(getattr(utils, func_name)):
                return False, f"Not callable: utils.{func_name}"
        
        return True, "All utils functions accessible"
    except Exception as e:
        return False, f"Error testing utils: {str(e)}"


def test_importers_package():
    """Test importers package is properly wired"""
    try:
        import importers
        
        # Check registry functions
        required_attrs = ['BROKER_REGISTRY', 'register_broker', 'get_importer', 'list_brokers']
        
        for attr in required_attrs:
            if not hasattr(importers, attr):
                return False, f"Missing attribute: importers.{attr}"
        
        # Check all brokers are registered
        expected_brokers = ['ibkr', 'schwab', 'robinhood', 'webull']
        registered_brokers = importers.list_brokers()
        
        for broker in expected_brokers:
            if broker not in registered_brokers:
                return False, f"Broker not registered: {broker}"
            
            importer = importers.get_importer(broker)
            if importer is None:
                return False, f"Cannot get importer for: {broker}"
        
        return True, f"All brokers registered: {registered_brokers}"
    except Exception as e:
        return False, f"Error testing importers: {str(e)}"


def test_base_importer():
    """Test BaseImporter class"""
    try:
        from importers.base_importer import BaseImporter
        
        required_methods = [
            'detect_format', 'parse_csv', 'validate_trade',
            'get_broker_name', 'get_supported_formats', 'get_sample_mapping'
        ]
        
        for method in required_methods:
            if not hasattr(BaseImporter, method):
                return False, f"Missing method: BaseImporter.{method}"
        
        return True, f"BaseImporter has all {len(required_methods)} required methods"
    except Exception as e:
        return False, f"Error testing BaseImporter: {str(e)}"


def main():
    """Main test execution"""
    print("=" * 70)
    print("Python Import and Accessibility Test")
    print("=" * 70)
    
    # Get base directory
    base_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(base_dir)
    
    # List of all Python files to test
    python_files = [
        'utils.py',
        'parse_trades.py',
        'import_csv.py',
        'generate_analytics.py',
        'generate_charts.py',
        'generate_index.py',
        'generate_summaries.py',
        'generate_trade_pages.py',
        'generate_week_summaries.py',
        'generate_books_index.py',
        'generate_notes_index.py',
        'update_homepage.py',
        'export_csv.py',
        'normalize_schema.py',
        'attach_media.py',
        'navbar_template.py',
        'importers/__init__.py',
        'importers/base_importer.py',
        'importers/ibkr.py',
        'importers/schwab.py',
        'importers/robinhood.py',
        'importers/webull.py'
    ]
    
    # Test 1: Import all Python files
    print("\n[Test 1] Testing all Python file imports...")
    print("-" * 70)
    
    failed_imports = []
    for file_path in python_files:
        success, message = test_file_import(file_path, base_dir)
        status = "✓" if success else "✗"
        print(f"{status} {file_path}")
        
        if not success:
            failed_imports.append(message)
    
    # Test 2: Test utils functions
    print("\n[Test 2] Testing utils.py functions...")
    print("-" * 70)
    success, message = test_utils_functions()
    status = "✓" if success else "✗"
    print(f"{status} {message}")
    if not success:
        failed_imports.append(message)
    
    # Test 3: Test importers package
    print("\n[Test 3] Testing importers package...")
    print("-" * 70)
    success, message = test_importers_package()
    status = "✓" if success else "✗"
    print(f"{status} {message}")
    if not success:
        failed_imports.append(message)
    
    # Test 4: Test BaseImporter
    print("\n[Test 4] Testing BaseImporter class...")
    print("-" * 70)
    success, message = test_base_importer()
    status = "✓" if success else "✗"
    print(f"{status} {message}")
    if not success:
        failed_imports.append(message)
    
    # Print results
    print("\n" + "=" * 70)
    print("TEST RESULTS")
    print("=" * 70)
    successful_imports = sum(1 for f in python_files if test_file_import(f, base_dir)[0])
    print(f"Total files tested: {len(python_files)}")
    print(f"Successful imports: {successful_imports}")
    print(f"Failed tests: {len(failed_imports)}")
    
    if failed_imports:
        print("\n⚠️  Failed tests:")
        for i, failure in enumerate(failed_imports, 1):
            print(f"  {i}. {failure}")
        print("\n" + "=" * 70)
        sys.exit(1)
    else:
        print("\n" + "=" * 70)
        print("✓✓✓ ALL TESTS PASSED! ✓✓✓")
        print("=" * 70)
        print("✓ All Python files are properly wired")
        print("✓ Global awareness and accessibility verified")
        print("✓ No silent failures detected")
        print("=" * 70)
        sys.exit(0)


if __name__ == "__main__":
    main()

File: system/scripts/test_imports.js
#!/usr/bin/env node
/**
 * Test Imports Script
 * Validates that all JavaScript files are properly wired with global awareness and accessibility
 * Tests for silent failures in imports and functions
 * 
 * This is a comprehensive JavaScript translation of test_imports.py with full feature parity.
 */

const path = require('path');
const fs = require('fs');

// Constants
const ERROR_MESSAGE_MAX_LENGTH = 100;

/**
 * Test if a JavaScript file can be imported without errors
 * 
 * Python equivalent: test_file_import(file_path, base_dir)
 * Uses: importlib.util for dynamic imports
 * Note: JavaScript uses require() for dynamic imports
 * 
 * @param {string} filePath - Path to JavaScript file relative to base_dir
 * @param {string} baseDir - Base directory containing the JavaScript files
 * @returns {Object} Object with {success: boolean, message: string}
 */
function testFileImport(filePath, baseDir) {
    try {
        const fullPath = path.join(baseDir, filePath);
        
        if (!fs.existsSync(fullPath)) {
            return { success: false, message: `File not found: ${filePath}` };
        }
        
        // Try to require the module
        const module = require(fullPath);
        
        return { success: true, message: `Successfully imported ${filePath}` };
        
    } catch (error) {
        const errorMsg = error.message.substring(0, ERROR_MESSAGE_MAX_LENGTH);
        return { success: false, message: `Error importing ${filePath}: ${errorMsg}` };
    }
}

/**
 * Test utils.js functions are accessible
 * 
 * Python equivalent: test_utils_functions()
 */
function testUtilsFunctions() {
    try {
        const utils = require('./utils');
        
        const requiredFunctions = ['loadTradesIndex', 'loadAccountConfig'];
        
        for (const funcName of requiredFunctions) {
            if (!(funcName in utils)) {
                return { success: false, message: `Missing function: utils.${funcName}` };
            }
            if (typeof utils[funcName] !== 'function') {
                return { success: false, message: `Not callable: utils.${funcName}` };
            }
        }
        
        return { success: true, message: 'All utils functions accessible' };
    } catch (error) {
        return { success: false, message: `Error testing utils: ${error.message}` };
    }
}

/**
 * Test importers package is properly wired
 * 
 * Python equivalent: test_importers_package()
 */
function testImportersPackage() {
    try {
        const importers = require('./importers/index');
        
        // Check registry functions
        const requiredAttrs = ['BROKER_REGISTRY', 'registerBroker', 'getImporter', 'listBrokers'];
        
        for (const attr of requiredAttrs) {
            if (!(attr in importers)) {
                return { success: false, message: `Missing attribute: importers.${attr}` };
            }
        }
        
        // Check all brokers are registered
        const expectedBrokers = ['ibkr', 'schwab', 'robinhood', 'webull'];
        const registeredBrokers = importers.listBrokers();
        
        for (const broker of expectedBrokers) {
            if (!registeredBrokers.includes(broker)) {
                return { success: false, message: `Broker not registered: ${broker}` };
            }
            
            const importer = importers.getImporter(broker);
            if (importer === null) {
                return { success: false, message: `Cannot get importer for: ${broker}` };
            }
        }
        
        return { success: true, message: `All brokers registered: ${registeredBrokers.join(', ')}` };
    } catch (error) {
        return { success: false, message: `Error testing importers: ${error.message}` };
    }
}

/**
 * Test BaseImporter class
 * 
 * Python equivalent: test_base_importer()
 */
function testBaseImporter() {
    try {
        const { BaseImporter } = require('./importers/base_importer');
        
        const requiredMethods = [
            'detectFormat', 'parseCsv', 'validateTrade',
            'getBrokerName', 'getSupportedFormats', 'getSampleMapping'
        ];
        
        for (const method of requiredMethods) {
            if (!(method in BaseImporter.prototype)) {
                return { success: false, message: `Missing method: BaseImporter.${method}` };
            }
        }
        
        return { success: true, message: `BaseImporter has all ${requiredMethods.length} required methods` };
    } catch (error) {
        return { success: false, message: `Error testing BaseImporter: ${error.message}` };
    }
}

/**
 * Main test execution
 * 
 * Python equivalent: main()
 */
function main() {
    console.log('='.repeat(70));
    console.log('JavaScript Import and Accessibility Test');
    console.log('='.repeat(70));
    
    // Get base directory
    const baseDir = __dirname;
    process.chdir(baseDir);
    
    // List of all JavaScript files to test
    const jsFiles = [
        'utils.js',
        'globals_utils.js',
        'navbar_template.js',
        'parse_trades.js',
        'import_csv.js',
        'export_csv.js',
        'generate_analytics.js',
        'generate_charts.js',
        'generate_index.js',
        'generate_summaries.js',
        'generate_trade_pages.js',
        'generate_week_summaries.js',
        'generate_books_index.js',
        'generate_notes_index.js',
        'update_homepage.js',
        'normalize_schema.js',
        'attach_media.js',
        'importers/index.js',
        'importers/base_importer.js',
        'importers/ibkr.js',
        'importers/schwab.js',
        'importers/robinhood.js',
        'importers/webull.js'
    ];
    
    // Test 1: Import all JavaScript files
    console.log('\n[Test 1] Testing all JavaScript file imports...');
    console.log('-'.repeat(70));
    
    const failedTests = [];
    for (const filePath of jsFiles) {
        const { success, message } = testFileImport(filePath, baseDir);
        const status = success ? '✓' : '✗';
        console.log(`${status} ${filePath}`);
        
        if (!success) {
            failedTests.push(message);
        }
    }
    
    // Test 2: Test utils functions
    console.log('\n[Test 2] Testing utils.js functions...');
    console.log('-'.repeat(70));
    let result = testUtilsFunctions();
    let status = result.success ? '✓' : '✗';
    console.log(`${status} ${result.message}`);
    if (!result.success) {
        failedTests.push(result.message);
    }
    
    // Test 3: Test importers package (skip if not implemented yet)
    console.log('\n[Test 3] Testing importers package...');
    console.log('-'.repeat(70));
    result = testImportersPackage();
    status = result.success ? '✓' : '✗';
    console.log(`${status} ${result.message}`);
    if (!result.success) {
        failedTests.push(result.message);
    }
    
    // Test 4: Test BaseImporter (skip if not implemented yet)
    console.log('\n[Test 4] Testing BaseImporter class...');
    console.log('-'.repeat(70));
    result = testBaseImporter();
    status = result.success ? '✗' : '✗';
    console.log(`${status} ${result.message}`);
    if (!result.success) {
        failedTests.push(result.message);
    }
    
    // Print results
    console.log('\n' + '='.repeat(70));
    console.log('TEST RESULTS');
    console.log('='.repeat(70));
    const successfulImports = jsFiles.filter(f => testFileImport(f, baseDir).success).length;
    console.log(`Total files tested: ${jsFiles.length}`);
    console.log(`Successful imports: ${successfulImports}`);
    console.log(`Failed tests: ${failedTests.length}`);
    
    if (failedTests.length > 0) {
        console.log('\n⚠️  Failed tests:');
        failedTests.forEach((failure, i) => {
            console.log(`  ${i + 1}. ${failure}`);
        });
        console.log('\n' + '='.repeat(70));
        process.exit(1);
    } else {
        console.log('\n' + '='.repeat(70));
        console.log('✓✓✓ ALL TESTS PASSED! ✓✓✓');
        console.log('='.repeat(70));
        console.log('✓ All JavaScript files are properly wired');
        console.log('✓ Global awareness and accessibility verified');
        console.log('✓ No silent failures detected');
        console.log('='.repeat(70));
        process.exit(0);
    }
}

// Run main if executed directly
if (require.main === module) {
    main();
}

module.exports = {
    testFileImport,
    testUtilsFunctions,
    testImportersPackage,
    testBaseImporter,
    main
};
